File: src/id.ts
import {
  childrenIJ as childrenIJS2,
  contains as containsS2,
  face as faceS2,
  fromFace as fromFaceS2,
  isFace as isFaceS2,
  level as levelS2,
  parent as parentS2,
  toIJ as toIJS2,
} from './s2/s2CellID';
import {
  children as childrenWM,
  contains as containsWM,
  fromID as fromIDWM,
  isFace as isFaceWM,
  level as levelWM,
  parent as parentWM,
  toID as toIDWM,
  toIJ as toIJWM,
} from './wm/mercID';

import type { Face, Projection } from './';

/**
 * @param proj - the projection of choice
 * @param id - the tile ID
 * @returns - the tile broken into face-zoom-i-j
 */
export function fromID(
  proj: Projection,
  id: bigint,
): [face: Face, zoom: number, i: number, j: number] {
  if (proj === 'S2') {
    const zoom = levelS2(id);
    const [face, i, j] = toIJS2(id, zoom);
    return [face, zoom, i, j];
  } else return [0, ...fromIDWM(id)];
}

/**
 * Given the projection, get the parent tile of the input ID
 * @param proj - the projection
 * @param id - the input tile ID
 * @returns - the tile ID of the parent
 */
export function parent(proj: Projection, id: bigint): bigint {
  if (proj === 'S2') return parentS2(id);
  else return parentWM(id);
}

/**
 * Given the projection, get the children tiles given the face-zoom-i-j.
 * @param proj - the projection
 * @param face - the face
 * @param zoom - the zoom
 * @param i - x position in tile space
 * @param j - y position in tile space
 * @returns - the tile IDs of the children
 */
export function childrenIJ(
  proj: Projection,
  face: Face,
  zoom: number,
  i: number,
  j: number,
): [blID: bigint, brID: bigint, tlID: bigint, trID: bigint] {
  if (proj === 'S2') return childrenIJS2(face, zoom, i, j);
  else return childrenWM(toIDWM(zoom, i, j));
}

/**
 * Given the projection, check if the tile ID is of the highest zoom or not.
 * @param proj - the projection
 * @param id - the input tile ID
 * @returns - true if the tile ID is of the highest zoom
 */
export function isFace(proj: Projection, id: bigint): boolean {
  if (proj === 'S2') return isFaceS2(id);
  else return isFaceWM(id);
}

/**
 * Given the projection, get the face of the input ID. If not S2, gaurenteed to be 0.
 * @param proj - the projection
 * @param id - the input tile ID
 * @returns - the tile ID of the parent
 */
export function face(proj: Projection, id: bigint): Face {
  if (proj === 'S2') return faceS2(id);
  else return 0 as Face;
}

/**
 * Given the projection, get the ID of a requested face. If not S2, gaurenteed to be 0.
 * @param proj - the projection
 * @param face - the requested face
 * @returns - the ID of the requested face
 */
export function fromFace(proj: Projection, face: Face): bigint {
  if (proj === 'S2') return fromFaceS2(face);
  else return 0n;
}

/**
 * Given the projection and id, return the face/zoom with the i-j coordinate pair
 * @param proj - the projection
 * @param id - the input tile ID
 * @param level - the requested zoom
 * @returns - the ID of the requested zoom
 */
export function toIJ(
  proj: Projection,
  id: bigint,
  level?: number | bigint,
): [faceOrZoom: number, i: number, j: number] {
  if (proj === 'S2') return toIJS2(id, level).slice(0, 3) as [face: number, i: number, j: number];
  else return toIJWM(id, level);
}

/**
 * Given the projection and tile ID, get the zoom of said tile.
 * @param proj - the projection
 * @param id - the input tile ID
 * @returns - the zoom of the tile
 */
export function level(proj: Projection, id: bigint): number {
  if (proj === 'S2') return levelS2(id);
  else return levelWM(id);
}

/**
 * Given the projection, check if the "child" exists inside the "parent"'s bounds or not.
 * @param proj - the projection
 * @param parentID - the parent
 * @param childID - the child
 * @returns - true if the "child" exists inside the "parent"'s bounds
 */
export function contains(proj: Projection, parentID: bigint, childID: bigint): boolean {
  if (proj === 'S2') return containsS2(parentID, childID);
  else return containsWM(parentID, childID);
}

File: rust/id.rs
use crate::s2::S2CellId;
use crate::wm::WMCellId;

/// Cell ID works with both S2 and WM with a common interface
pub enum CellId {
    /// S2 Cell ID
    S2(S2CellId),
    /// WM Cell ID
    WM(WMCellId),
}
impl CellId {
    /// Return the subdivision level of the cell
    pub fn level(&self) -> u8 {
        match self {
            CellId::S2(id) => id.level(),
            CellId::WM(id) => id.level(),
        }
    }

    /// Which cube face this cell belongs to, in the range 0..5 for S2, always 0 for WM
    pub fn face(&self) -> u8 {
        match self {
            CellId::S2(id) => id.face(),
            CellId::WM(_) => 0,
        }
    }

    /// Given an S2CellID, check if it is a Face Cell (zoom 0).
    pub fn is_face(&self) -> bool {
        match self {
            CellId::S2(id) => id.is_face(),
            CellId::WM(id) => id.is_face(),
        }
    }

    /// Return the cell corresponding to a given S2 cube face.
    /// WM returns 0
    pub fn from_face(face: u8) -> CellId {
        match face {
            0 => CellId::S2(S2CellId::from_face(face)),
            _ => CellId::WM(WMCellId::new(0)),
        }
    }

    /// Given an ID, get the parent quad tile
    pub fn parent(&self) -> CellId {
        match self {
            CellId::S2(id) => CellId::S2(id.parent(None)),
            CellId::WM(id) => CellId::WM(id.parent()),
        }
    }

    /// Check if the first CellID contains the second (its a subset)
    pub fn contains(&self, other: &CellId) -> bool {
        match (self, other) {
            (CellId::S2(id1), CellId::S2(id2)) => id1.contains(id2),
            (CellId::WM(id1), CellId::WM(id2)) => id1.contains(id2),
            _ => false,
        }
    }

    /// Given the projection and id, return the face/zoom with the i-j coordinate pair
    /// Returns (face/zoom, i, j)
    pub fn to_zoom_ij(&self, level: Option<u8>) -> (u8, u32, u32) {
        match self {
            CellId::S2(id) => id.to_zoom_ij(level),
            CellId::WM(id) => id.to_zoom_ij(level),
        }
    }

    /// Given the projection, get the children tile IDs
    pub fn children_ij(&self, face: u8, zoom: u8, i: u32, j: u32) -> [CellId; 4] {
        match self {
            CellId::S2(id) => id.children_ij(face, zoom, i, j).map(CellId::S2),
            CellId::WM(id) => id.children().map(CellId::WM),
        }
    }

    /// Decode the CellId back into (face, zoom, x, y) components.
    pub fn to_face_zoom_xy(&self) -> (u8, u8, u32, u32) {
        match self {
            CellId::S2(id) => {
                let zoom = id.level();
                let (face, x, y) = id.to_zoom_ij(None);
                (face, zoom, x, y)
            }
            CellId::WM(id) => {
                let (zoom, x, y) = id.to_zoom_xy();
                (0, zoom, x, y)
            }
        }
    }
}
